eras/shelley/impl/src/Cardano/Ledger/Shelley/AdaPots.hs

```diff
@@ -72,6 +92,87 @@ totalAdaES cs =
         reservesAdaPot,
         rewardsAdaPot,
         utxoAdaPot,
+        -- keyDepositAdaPot,  -- We don't count these two, as their
+        -- poolDepositAdaPot, -- sum is always depositsAdaPot
         depositsAdaPot,
         feesAdaPot
       } = totalAdaPotsES cs
+
+-- =============================================
+-- Produced and Consumed are specialized AdaPots
+-- relative to the actions of a TxBody
+
+-- | Itemizing what is consumed by a transaction
+data Consumed = Consumed
+  {conInputs :: !Coin, conRefunds :: !Coin, conWithdrawals :: !Coin}
+
+instance Show Consumed where
+  show (Consumed (Coin i) (Coin r) (Coin w)) =
+    "Consumed(Inputs " ++ show i ++ ", Refunds " ++ show r ++ ", Withdrawals " ++ show w ++ ") = " ++ show (i + r + w)
+
+-- | Itemizing what is Produced by a transaction
+data Produced = Produced
+  {proOutputs :: !Coin, proFees :: !Coin, proDeposits :: !Coin}
+
+instance Show Produced where
+  show (Produced (Coin out) (Coin f) (Coin d)) =
+    "Produced(Outputs " ++ show out ++ ", Fees " ++ show f ++ " Deposits " ++ show d ++ ") = " ++ show (out + f + d)
+
+-- =========================
+
+-- | Compute the Coin part of what is consumed by a TxBody, itemized as a 'Consume'
+consumedTxBody ::
+  ( HasField "_keyDeposit" pp Coin,
+    ShelleyEraTxBody era
+  ) =>
+  TxBody era ->
+  pp ->
+  DPState (EraCrypto era) ->
+  UTxO era ->
+  Consumed
+consumedTxBody txBody pp dpstate (UTxO u) = Consumed {conInputs = i, conRefunds = r, conWithdrawals = w}
+  where
+    i = coinBalance (UTxO (Map.restrictKeys u (txBody ^. inputsTxBodyL)))
+    r = keyTxRefunds pp dpstate txBody
+    w = fold . unWdrl $ txBody ^. wdrlsTxBodyL
+
+-- | Compute the Coin part of what is produced by a TxBody, itemized as a 'Produced'
+producedTxBody ::
+  ( ShelleyEraTxBody era,
+    HasField "_keyDeposit" pp Coin,
+    HasField "_poolDeposit" pp Coin
+  ) =>
+  TxBody era ->
+  pp ->
+  DPState (EraCrypto era) ->
+  Produced
+producedTxBody txBody pp dpstate = Produced {proOutputs = out, proFees = f, proDeposits = d}
+  where
+    out = coinBalance (txouts txBody)
+    f = txBody ^. feeTxBodyL
+    d = totalTxDeposits pp dpstate txBody
+
+-- ==============================================
+-- Compare two AdaPots, item by item
+
+pad :: Int -> String -> String
+pad n x = if m < n then x ++ [const ' ' i | i <- [1 .. n - m]] else x
+  where
+    m = length x
```

Comment:
I feel that this would be easier to read. `pad n x = x ++ replicate (n - length x) ' '`